### PR TITLE
Support UniFormer

### PR DESCRIPTION
**Pull request by thinksoso**
## Support Uniformer
- [x] build uniformer model
- [x] convert pretrained weight
- [x] inference test on imagenet and update model_zoo small_plus
- [x] update docs
- [x] update changelog 
- [x] pytorch speed comparison

### CODE REVIEW

File: flowvision/models/uniformer.py [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+
+        self.qkv = nn.Linear(dim, dim * 3, bias=qkv_bias)
+        self.attn_drop = nn.Dropout(attn_drop)
+        self.proj = nn.Linear(dim, dim)
+        self.proj_drop = nn.Dropout(proj_drop)
+
+    def forward(self, x):
+        B, N, C = x.shape
+        qkv = self.qkv(x).reshape(B, N, 3, self.num_heads, C // self.num_heads).permute(2, 0, 3, 1, 4)
+        q, k, v = qkv[0], qkv[1], qkv[2]   # make flowscript happy (cannot use tensor as tuple)
```
Comment (kaijieshi7) [inline flowvision/models/uniformer.py:84]: 这个原来是 make torchscript happy吧，暂时好像没听说有flowscript，感觉直接删除注释就行了

File: flowvision/models/uniformer.py [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+    Args:
+        pretrained (bool): Whether to download the pre-trained model on ImageNet. Default: ``False``
+        progress (bool): If True, displays a progress bar of the download to stderr. Default: ``True``
+
+    For example:
+
+    .. code-block:: python
+
+        >>> import flowvision
+        >>> pvt_tiny = flowvision.models.uniformer_small(pretrained=False, progress=True)
```
Comment (kaijieshi7) [inline flowvision/models/uniformer.py:369]: 这个改一下

File: flowvision/models/uniformer.py [hunk 3]
```diff
@@ -0,0 +1,462 @@
...
+    Args:
+        pretrained (bool): Whether to download the pre-trained model on ImageNet. Default: ``False``
+        progress (bool): If True, displays a progress bar of the download to stderr. Default: ``True``
+
+    For example:
+
+    .. code-block:: python
+
+        >>> import flowvision
+        >>> pvt_tiny = flowvision.models.uniformer_small_plus(pretrained=False, progress=True)
```
Comment (kaijieshi7) [inline flowvision/models/uniformer.py:397]: 改

File: flowvision/models/uniformer.py [hunk 4]
```diff
@@ -0,0 +1,462 @@
...
+    Args:
+        pretrained (bool): Whether to download the pre-trained model on ImageNet. Default: ``False``
+        progress (bool): If True, displays a progress bar of the download to stderr. Default: ``True``
+
+    For example:
+
+    .. code-block:: python
+
+        >>> import flowvision
+        >>> pvt_tiny = flowvision.models.uniformer_base(pretrained=False, progress=True)
```
Comment (kaijieshi7) [inline flowvision/models/uniformer.py:425]: 改

File: flowvision/models/uniformer.py [hunk 5]
```diff
@@ -0,0 +1,462 @@
...
+    Args:
+        pretrained (bool): Whether to download the pre-trained model on ImageNet. Default: ``False``
+        progress (bool): If True, displays a progress bar of the download to stderr. Default: ``True``
+
+    For example:
+
+    .. code-block:: python
+
+        >>> import flowvision
+        >>> pvt_tiny = flowvision.models.uniformer_base_ls(pretrained=False, progress=True)
```
Comment (kaijieshi7) [inline flowvision/models/uniformer.py:453]: 改

File: flowvision/models/uniformer.py (hunk 1 above)
Comment (rentainhe) [reply flowvision/models/uniformer.py:84]: 这里注释删了吧~没有flowscript目前

**Comment by thinksoso** [comment]
eval 结果跟论文对不上的问题已经向官方提了 issue https://github.com/Sense-X/UniFormer/issues/16

**Comment by thinksoso** [comment]
### 测速结果
OneFlow uniformer_base time: 200.5ms (= 20045.4ms / 100, input_shape=[16, 3, 224, 224])
PyTorch uniformer_base time: 180.3ms (= 18030.1ms / 100, input_shape=[16, 3, 224, 224])
❌ Relative speed: 0.90 (= 180.3ms / 200.5ms)

PyTorch uniformer_base_ls time: 188.0ms (= 18803.9ms / 100, input_shape=[16, 3, 224, 224])
❌ Relative speed: 0.92 (= 188.0ms / 205.4ms)

OneFlow uniformer_small time: 83.6ms (= 8361.4ms / 100, input_shape=[16, 3, 224, 224])
PyTorch uniformer_small time: 102.1ms (= 10210.4ms / 100, input_shape=[16, 3, 224, 224])
✔️ Relative speed: 1.22 (= 102.1ms / 83.6ms)

OneFlow uniformer_small_plus time: 99.2ms (= 9921.4ms / 100, input_shape=[16, 3, 224, 224])
PyTorch uniformer_small_plus time: 108.9ms (= 10885.4ms / 100, input_shape=[16, 3, 224, 224])
✔️ Relative speed: 1.10 (= 108.9ms / 99.2ms)

OneFlow uniformer_base time: 257.9ms (= 25788.3ms / 100, input_shape=[16, 3, 224, 224], ddp, world size=2)
PyTorch uniformer_base time: 265.4ms (= 26543.0ms / 100, input_shape=[16, 3, 224, 224], ddp, world size=2)
✔️ Relative speed: 1.03 (= 265.4ms / 257.9ms)

✔️ Relative speed: 0.99 (= 273.9ms / 275.5ms)

OneFlow uniformer_small time: 121.7ms (= 12169.2ms / 100, input_shape=[16, 3, 224, 224], ddp, world size=2)
PyTorch uniformer_small time: 123.5ms (= 12352.7ms / 100, input_shape=[16, 3, 224, 224], ddp, world size=2)
✔️ Relative speed: 1.02 (= 123.5ms / 121.7ms)

OneFlow uniformer_small_plus time: 139.7ms (= 13967.8ms / 100, input_shape=[16, 3, 224, 224], ddp, world size=2)
PyTorch uniformer_small_plus time: 137.3ms (= 13732.3ms / 100, input_shape=[16, 3, 224, 224], ddp, world size=2)
❌ Relative speed: 0.98 (= 137.3ms / 139.7ms)

File: ci/check/run_speed_test.sh [hunk 6]
```diff
@@ -25,8 +25,11 @@ python3 $COMPARE_SCRIPT_PATH $MODELS_ROOT/convnext.py convnext_tiny_224 16x3x224
...
-python3 $COMPARE_SCRIPT_PATH $MODELS_ROOT/res_mlp.py resmlp_24_224 16x3x224x224 --no-show-memory --times $TIMES | check_relative_speed 1.01 | write_to_file_and_print
+python3 $COMPARE_SCRIPT_PATH $MODELS_ROOT/uniformer.py uniformer_base 16x3x224x224 --no-show-memory --times $TIMES | check_relative_speed 1.01 | write_to_file_and_print
+python3 $COMPARE_SCRIPT_PATH $MODELS_ROOT/uniformer.py uniformer_base_ls 16x3x224x224 --no-show-memory --times $TIMES | check_relative_speed 1.01 | write_to_file_and_print
+python3 $COMPARE_SCRIPT_PATH $MODELS_ROOT/uniformer.py uniformer_small 16x3x224x224 --no-show-memory --times $TIMES | check_relative_speed 1.01 | write_to_file_and_print
+python3 $COMPARE_SCRIPT_PATH $MODELS_ROOT/uniformer.py uniformer_small_plus 16x3x224x224 --no-show-memory --times $TIMES | check_relative_speed 1.01 | write_to_file_and_print
+python3 $COMPARE_SCRIPT_PATH $MODELS_ROOT/res_mlp.py resmlp_12_224 16x3x224x224 --no-show-memory --times $TIMES | check_relative_speed 1.01 | write_to_file_and_print # python3 $COMPARE_SCRIPT_PATH $MODELS_ROOT/res_mlp.py resmlp_24_224 16x3x224x224 --no-show-memory --times $TIMES | check_relative_speed 1.01 | write_to_file_and_print
```
Comment (rentainhe) [inline ci/check/run_speed_test.sh:32]: 这儿是不是有一些注释的内容，要删了或者update一下

File: flowvision/models/uniformer.py [hunk 7]
```diff
@@ -0,0 +1,463 @@
...
+from flowvision.layers import trunc_normal_, DropPath
+from .helpers import to_2tuple
+from .registry import ModelCreator
+from .utils import load_state_dict_from_url
+
+model_urls = {
+    "uniformer_base": "https://oneflow-public.oss-cn-beijing.aliyuncs.com/model_zoo/flowvision/classification/UniFormer/uniformer_base_oneflow.zip",
+    "uniformer_base_ls": "https://oneflow-public.oss-cn-beijing.aliyuncs.com/model_zoo/flowvision/classification/UniFormer/uniformer_base_ls_oneflow.zip",
+    "uniformer_small": "https://oneflow-public.oss-cn-beijing.aliyuncs.com/model_zoo/flowvision/classification/UniFormer/uniformer_small_oneflow.zip",
+    # "uniformer_small_plus": "https://oneflow-public.oss-cn-beijing.aliyuncs.com/model_zoo/flowvision/classification/UniFormer/uniformer_small_plus_oneflow.zip",
```
Comment (rentainhe) [inline flowvision/models/uniformer.py:22]: 不用注释，将这里的链接设置为`None`就行了
```python
uniformer_small_plus: None
```

File: flowvision/models/uniformer.py [hunk 8]
```diff
@@ -0,0 +1,463 @@
...
+    def _init_weights(self, m):
+        if isinstance(m, nn.Linear):
+            trunc_normal_(m.weight, std=.02)
+            if isinstance(m, nn.Linear) and m.bias is not None:
+                nn.init.constant_(m.bias, 0)
+        elif isinstance(m, nn.LayerNorm):
+            nn.init.constant_(m.bias, 0)
+            nn.init.constant_(m.weight, 1.0)
+
+    # @torch.jit.ignore
```
Comment (rentainhe) [inline flowvision/models/uniformer.py:308]: 和torch有关的注释删了~

File: results/results_imagenet.md [hunk 9]
```diff
@@ -143,6 +143,9 @@
 | convnext_iso_small_224        | 79.742 | 20.258 | 94.798 | 5.208 | 22.3M  | 224 | 0.875 | bicubic |
 | convnext_iso_base_224         | 81.818 | 18.182 | 95.674 | 4.326 | 87.1M  | 224 | 0.875 | bicubic |
 | convnext_iso_large_224        | 82.576 | 17.424 | 95.956 | 4.044 | 306.0M | 224 | 0.875 | bicubic |
+| uniformer_base                | 83.814 | 16.186 | 96.738 | 3.262 | 50.0M | 224 | 0.900 | bicubic |
+| uniformer_base_ls             | 83.952 | 16.048 | 96.728 | 3.272 | 50.0M | 224 | 0.900 | bicubic |
+| uniformer_small               | 82.958 | 17.042 | 96.290 | 3.710 | 22.0M | 224 | 0.900 | bicubic |
```
Comment (rentainhe) [inline results/results_imagenet.md:148]: small模型的测试结果放到base前面吧，按`small` -> `base`的顺序

File: flowvision/models/uniformer.py [hunk 10]
```diff
@@ -0,0 +1,463 @@
...
+    if pretrained:
+        state_dict = load_state_dict_from_url(model_urls[arch], progress=progress)
+        model.load_state_dict(state_dict)
+    return model
+
+
+@ModelCreator.register_model
+def uniformer_small(pretrained=False, progress=True, **kwargs):
+    """
+    Constructs the UniFormer-small model.
```
Comment (rentainhe) [inline flowvision/models/uniformer.py:355]: 这儿最好两个字母都大写`Uniformer-small` -> `UniFormer-Small`

File: README.md [hunk 11]
```diff
@@ -71,6 +71,7 @@ pip install flowvision==0.1.0
             <li>Swin Transformer</li>
             <li>CSwin Transformer</li>
             <li>CrossFormer</li>
+            <li>UniFormer<li>
```
Comment (rentainhe) [inline README.md:74]: UniFormer和RegionViT好像是一个会议的，放到RegionViT之后吧

**Comment by rentainhe** [comment]
@kaijieshi7 开杰这边double review一下，我没有问题了~

File: flowvision/models/uniformer.py [hunk 12]
```diff
@@ -0,0 +1,461 @@
...
+            head_dim (int): head dimension
+            mlp_ratio (int): ratio of mlp hidden dim to embedding dim
+            qkv_bias (bool): enable bias for qkv if True
+            qk_scale (float): override default qk scale of head_dim ** -0.5 if set
+            representation_size (Optional[int]): enable and set representation layer (pre-logits) to this value if set
+            drop_rate (float): dropout rate
+            attn_drop_rate (float): attention dropout rate
+            drop_path_rate (float): stochastic depth rate
+            norm_layer (nn.Module): normalization layer
+            conv_stem (bool): whether use overlapped patch stem
```
Comment (kaijieshi7) [inline flowvision/models/uniformer.py:237]: 是不是whether to do sth.

**Comment by Andy1621** [comment]
抱歉，[issue](https://github.com/Sense-X/UniFormer/issues/16)已回复，精度没对上是模型代码有个参数写错了，现在的model我都测了一遍没有问题的

File: flowvision/models/uniformer.py (hunk 12 above)
Comment (Aries-chen) [reply flowvision/models/uniformer.py]: 一般加从句 或者 to do

**Review by Aries-chen — APPROVED** [review]

**Comment by Andy1621** [comment]
> ### 测速结果
> OneFlow uniformer_base time: 198.4ms (= 19835.6ms / 100, input_shape=[16, 3, 224, 224]) PyTorch uniformer_base time: 205.1ms (= 20506.7ms / 100, input_shape=[16, 3, 224, 224]) ✔️ Relative speed: 1.03 (= 205.1ms / 198.4ms)
> 
> PyTorch uniformer_base_ls time: 211.3ms (= 21126.4ms / 100, input_shape=[16, 3, 224, 224]) ✔️ Relative speed: 1.06 (= 211.3ms / 198.9ms)
> 
> OneFlow uniformer_small time: 83.6ms (= 8361.4ms / 100, input_shape=[16, 3, 224, 224]) PyTorch uniformer_small time: 102.1ms (= 10210.4ms / 100, input_shape=[16, 3, 224, 224]) ✔️ Relative speed: 1.22 (= 102.1ms / 83.6ms)
> 
> OneFlow uniformer_small_plus time: 99.2ms (= 9921.4ms / 100, input_shape=[16, 3, 224, 224]) PyTorch uniformer_small_plus time: 108.9ms (= 10885.4ms / 100, input_shape=[16, 3, 224, 224]) ✔️ Relative speed: 1.10 (= 108.9ms / 99.2ms)
> 
> OneFlow uniformer_base time: 257.9ms (= 25788.3ms / 100, input_shape=[16, 3, 224, 224], ddp, world size=2) PyTorch uniformer_base time: 265.4ms (= 26543.0ms / 100, input_shape=[16, 3, 224, 224], ddp, world size=2) ✔️ Relative speed: 1.03 (= 265.4ms / 257.9ms)
> 
> ✔️ Relative speed: 0.99 (= 273.9ms / 275.5ms)
> 
> OneFlow uniformer_small time: 121.7ms (= 12169.2ms / 100, input_shape=[16, 3, 224, 224], ddp, world size=2) PyTorch uniformer_small time: 123.5ms (= 12352.7ms / 100, input_shape=[16, 3, 224, 224], ddp, world size=2) ✔️ Relative speed: 1.02 (= 123.5ms / 121.7ms)
> 
> OneFlow uniformer_small_plus time: 139.7ms (= 13967.8ms / 100, input_shape=[16, 3, 224, 224], ddp, world size=2) PyTorch uniformer_small_plus time: 137.3ms (= 13732.3ms / 100, input_shape=[16, 3, 224, 224], ddp, world size=2) ❌ Relative speed: 0.98 (= 137.3ms / 139.7ms)

您好，不同模型的测速结果，oneflow官方有提供吗，对专门的测速比较挺感兴趣

**Comment by kaijieshi7** [comment]
https://github.com/Oneflow-Inc/vision/blob/main/ci/check/compare_speed_with_pytorch.py
在这里，最简单的是直接测试，这里的代码是转换引用的

**Pull request merged**